File: 2333-minimum-sum-of-squared-difference/2333-minimum-sum-of-squared-difference.cpp
class Solution {
public:
    int n;
    long long minSumSquareDiff(vector<int>& nums1, vector<int>& nums2, int k1, int k2) 
    {
        n=nums1.size();
        int target=k1+k2;
        vector<long long>nums;
        for(int i=0;i<n;i++)
        {
            nums.push_back(abs(nums1[i]-nums2[i]));
        }
        sort(nums.begin(),nums.end(),greater<int>());
        long long l=0;
        long long r=1e9;
        long long ans=-1;
        while(l<=r)
        {
            long long mid=(l+r)/2;
            long long sum=0;
            for(int i=0;i<n;i++)
            {
                sum+=max(0ll,nums[i]-mid);
            }
            if(sum<=target)
            {
                ans=mid;
                r=mid-1;
            }
            else
            {
                l=mid+1;
            }
        }
        
        long long result=0;
        for(int i=0;i<n;i++)
        {
            target-=max(0ll,nums[i]-ans);
            nums[i]=min(nums[i],ans);
        }
        for(int i=0;i<min(target,n);i++)
        {
            if(nums[i])
            nums[i]--;
        }
        for(int i=0;i<n;i++)
        {
            result+=(nums[i]*nums[i]);
        }
        return result;
    }
};

File: 2333-minimum-sum-of-squared-difference/2333-minimum-sum-of-squared-difference.py
class Solution:
    def minSumSquareDiff(self, nums1: List[int], nums2: List[int], k1: int, k2: int) -> int:
        n=len(nums1)
        target=k1+k2
        nums=[]
        for i in range(0,n):
            nums.append(abs(nums1[i]-nums2[i]))
        nums.sort(reverse=True)
        l=0
        r=10**9
        ans=-1
        while l<=r:
            mid=(l+r)//2
            count=0
            for i in nums:
                count+=max(0,i-mid)
            if count<=target:
                r=mid-1
                ans=mid
            else:
                l=mid+1
        for i in range(0,n):
            target-=max(0,nums[i]-ans)
            nums[i]=min(nums[i],ans)
        for i in range(0,n):
            if target>0 and nums[i]>0:
                nums[i]-=1
                target-=1
        result=0
        for i in nums:
            result+=(i*i)
        return result
            
        
        